Allow setting CBCentralManager initialization flags

JustusvonBrandt opened the pull request:
If an app does not want to use the bluetooth background mode, the `CBCentralManager` should not have the flags `CBConnectPeripheralOptionNotifyOnDisconnectionKey` and `CBConnectPeripheralOptionNotifyOnConnectionKey` set to `true`, since the notifications are not providing any value to the user. So I added a setter to be able to change the global variable called `standardConnectOptions` in the `Connection.swift` file.

JeremyChiang commented:
Thank you for the pull request. This issue has been on my radar as well. I would like to suggest a few more changes:

1. `var standardConnectOptions` currently reside in `Connection.swift`, that's just because it was convenient when getting started, but there is probably a better place to store that variable.
2. Not sure if `standardConnectOptions` should be a global variable or encapsulated, could discuss more, but I'm actually leaning towards `standardConnectOptions` being encapsulated, local, and specific to each instance of `Connection`
3. If we follow my train of thought, `standardConnectOptions` should be a default parameter, but also specifiable when creating a connection request.
4. We can still provide a Bluejay-level setter for `standardConnectOptions` that updates what goes into each `Connection` by default, but if so, we should add a getter as well for evaluation purposes.
5. We can also encapsulate the `CBConnectPeripheralOptionNotifyOnDisconnectionKey` and `CBConnectPeripheralOptionNotifyOnConnectionKey` constants by using an enum. That way they don't always need to import `CoreBluetooth`, and our documentation and comments can be more friendly than Apple's.

I think that's it for now, let me know if these ideas make sense. Cheers.

JustusvonBrandt commented:
Thanks for the detailed feedback. I also thought about scoping the connection options to the individual connections, so I took some time and reworked the pull request. 
Now each `Connection` is initialized with options, which are the default options from before if no others are set (to not break any existing implementations). To do this I also needed to give the `Scan` actions the options, since the `ScanAction.connect(...)` should also use the custom options. 

I also added the getter for the options and the enum for the `CoreBluetooth` keys. The enum has the advantage of being easy to use, but only the enum case supported keys are usable for the connections. So I also also added the key for the bluetooth powered off alert, since I think this is useful as well.

Let me know what you think about it.

JeremyChiang commented:
Hi @JustusvonBrandt thank you for the awesome work, it is looking really good. I am planning to test this PR next week when I get back to the office using some of our ongoing projects that rely on Bluejay heavily, so ETA for merge and next release is sometime before next Wednesday! 👍

JeremyChiang commented:
Hi, I just want to drop a quick update: I've been working on this PR since this morning, and I'm close to getting something out soon. I will post a more detailed description and explanations on the changes I've made later.

JeremyChiang commented:
Hi @JustusvonBrandt and @nbrooke, I've made some significant and hopefully positive changes to this PR. Here are the details:

## Overview
The changes have these goals in mind:
- Make the interface easier for new and existing Bluejay users
- Xcode auto-complete and inline documentation should unblock and be self-explanatory when they encounter either manager or peripheral options as they code

## Changes
- There was a mistake lumping `CBCentralManagerOptionShowPowerAlertKey` along with `CBConnectPeripheralOptionNotifyOnConnectionKey` and its disconnection counterpart.
  - Fix is to allow specifying `CBCentralManagerOptionShowPowerAlertKey` during `start`, and restrict connection options to the initializers of `Connection`

```
public func start(
    connectionObserver observer: ConnectionObserver? = nil,
    enableBluetoothAlert showBluetoothAlert: Bool = false,
    backgroundRestore restoreMode: BackgroundRestoreMode = .disable,
    coreBluetoothState: (manager: CBCentralManager, peripheral: CBPeripheral?)? = nil
    )
```
and for `Connection`
```
init(peripheral: CBPeripheral, manager: CBCentralManager, timeout: Timeout, connectionOptions: [ConnectionOption], callback: @escaping (ConnectionResult) -> Void) {
    self.state = .notStarted
        
    self.peripheral = peripheral
    self.manager = manager
    self.timeout = timeout
    self.connectionOptions = connectionOptions
    self.callback = callback
}
```
- Simplify the `ConnectionOptions` param, so the users only have predetermined and self-explanatory options and do not have to worry about creating dictionaries and CoreBluetooth constants.

Here's what the user experience might be like when you first need to use it or encounter it. Harder to get wrong, I hope.
![screen shot 2018-08-30 at 11 09 25 pm](https://user-images.githubusercontent.com/4675020/44895583-ce932380-aca9-11e8-8f52-6484a0403f04.png)

- Instead of specifying connection options when creating a `Scan` when you know you want to connect immediately after finding your peripheral, again, I think it's better to keep the connection options associated with the connection itself, so I moved the connection options from `Scan`'s initializer to the `ScanAction` enum instead:
```
/// Indicates whether a scan should continue, continue but blacklist the current discovery, stop, or stop and connect.
public enum ScanAction {
    /// Continue scanning.
    case `continue`
    /// Continue scanning, but don't discover the same peripheral in the current callback again within the same scan session.
    case blacklist
    /// Stop scanning.
    case stop
    /// Stop scanning, and connect to a discovery.
    case connect(ScanDiscovery, Timeout, ConnectionOptions, (ConnectionResult) -> Void)
}
```
- Last but not least, I don't think it is necessary to block setting and changing Bluejay's default connection options, as long as it is clear that it will only affect future new connection attempts. 
  - We also had 2 variables plus one setter function dealing with connection options in Bluejay, and that was kind of confusing, so I removed all of it and just kept 1 that you can both read and write:
```
/// Connection options to use for each new connection if the options are not specified at the creation of those connections.
    public var defaultConnectionOptions = ConnectionOptions.defaultOptions
```
  - Since each `Connection` holds a reference to the connection options given to them, if we need to debug or examine connection options, that state and value should come from the `Connection` objects anyways, not Bluejay

JustusvonBrandt commented:
Looks very good, much cleaner than what I did in the first place.

JeremyChiang commented:
Thanks @JustusvonBrandt, Nigel should be back next week and once we get a chance to review this again internally, we will release to branch `v0.7.0`.

nbrooke commented on Bluejay/Bluejay/Bluejay.swift:
```diff
@@ -133,11 +136,13 @@ public class Bluejay: NSObject {
     public func start(
         connectionObserver observer: ConnectionObserver? = nil,
+        enableBluetoothAlert showBluetoothAlert: Bool = false,
         backgroundRestore restoreMode: BackgroundRestoreMode = .disable,
```
This might be overthinking it, but it seems weird that in the case of the central manager options we are just directly including the properties that go into the options dictionary in the function call signature, even though it has basically the same sort of semantics as the ConnectionOptions, for which we are using a wrapper struct. Seems like if we wanted to be really consistent  enableBluetoothAlert and backgroundRestore should be in some "StartOptions" struct or something.

On the other hand, the main reason for the struct in connections is to enable setting the global defaults, which I don't think is as useful (maybe not at all useful) for initialization. If we didn't want to be able to do that global default, the connection options would probably be nicer as just parameters to the connect function as well. So it also seems bad to add extra abstraction here that is unnecessary and probably worse for users just to be consistent.

Not sure if I'm actually advocating for changing it or not, I think on balance I lean towards leaving it like it is now, but I'm unsure enough I thought it was worth raising as a possibility.

JeremyChiang replied:
I might not fully understand the problem raised here yet. Bluejay's `enableBluetoothAlert` param maps to CB's `CBCentralManagerOptionShowPowerAlertKey` and this pair is different from Bluejay's `ConnectionOptions` struct which maps to `CBConnectPeripheralOptionNotifyOnConnectionKey` or `CBConnectPeripheralOptionNotifyOnDisconnectionKey`. I think they do have different semantics, as we're comparing Bluetooth availability versus peripheral connection, so perhaps the distinction here is okay.

In fact, I think we might be more "correct" in the usage of CB by not specifying and dealing with neither `CBConnectPeripheralOptionNotifyOnConnectionKey` nor `CBConnectPeripheralOptionNotifyOnDisconnectionKey` when initializing the manager, as the scope at that point doesn't really involve any connections just yet. And if there are connections to restore, it should already have the options cached with the connection that' about to be restored.

nbrooke replied:
I'm not saying that the options should be combined in any way, or that we should be passing the connection options into the start struct. 

All I was pointing out is that we have one set of options (the connection options) that are passed in as a struct and then turned into a dictionary to pass to CoreBluetooth, and another set that are passed in as parameters to a function then turned into a dictionary object to pass to CB.

I'm speculating that maybe the signature for start should be something like:

```
struct StartOptions {
    var enableBluetoothAlert showBluetoothAlert: Bool = false,
    var backgroundRestore restoreMode: BackgroundRestoreMode = .disable
}

func start(connectionObserver observer: ConnectionObserver? = nil, options: StartOptions? = nil, coreBluetoothState: (manager: CBCentralManager, peripheral: CBPeripheral?)? = nil)
```

So that it follows the same approach as connection, the options that are going to be turned into the option dictionary to get passed to CoreBluetooth at manager initialization are wrapped up. In a struct.

Like I said, I'm not sure if this is actually better, it's just a little more consistent with the way we deal with options in connection.

In fact, taking it a little further, we should actually be restricted to EITHER specifying the start options or an existing CB state, since those two are actually mutually exclusive.

```
struct StartOptions {
    var enableBluetoothAlert showBluetoothAlert: Bool = false,
    var backgroundRestore restoreMode: BackgroundRestoreMode = .disable
}

enum StartMode {
    case create(StartOptions?)
    case useExisting(manager: CBCentralManager, peripheral: CBPeripheral?)
}

func start(connectionObserver observer: ConnectionObserver? = nil, mode: StartMode)

```

nbrooke commented on Bluejay/Bluejay/Bluejay.swift:
```diff
@@ -380,9 +385,15 @@ public class Bluejay: NSObject {
-    public func connect(_ peripheralIdentifier: PeripheralIdentifier, timeout: Timeout, completion: @escaping (ConnectionResult) -> Void) {
+    public func connect(
+        _ peripheralIdentifier: PeripheralIdentifier,
+        timeout: Timeout,
```
Another question in the "function arguments vs. struct members" domain: Should "timeout" be in the ConnectionOptions struct a well? 

It does make more sense the is some ways, and it could be useful to be able to default it as well, but that would be a breaking API change, and it doesn't totally fit the behind the scenes meaning of that struct (mapping to the CB connection options dictionary), since timeout is a Bluejay feature.

As with using a options struct to do initial configuration mentioned above, my gut feeling is that change probably isn't worth it, but again wanted to just raise it for consideration.

JeremyChiang replied:
Good point, this makes me feel maybe `ConnectionOptions` is just poorly named. Yes that does map to CB's terminology, but since we're abstracting it, we could name it anything we want to further our goal of making it clearer and self-explanatory to the user. For example, `WarningOptions`. That way, the distinction between `Timeout` and `WarningOptions` is clear, and it's also clearer that if you want to work with `WarningOptions`, you're going to deal with some kind of warning behaviours related to connection. Would this be better?

nbrooke replied:
I think 'WarningOptions' probably WOULD be a better name for it as it stands right now. I

 guess whether to not we change it depends on how we want to evolve it in the future. If we think that there aren't going to be any new options added, or if they were added we would want to deal with them separately rather than wrapping them up in the same struct. If so, i'd say we rename to 'WarningOptions'. 

If we think there will be more stuff going in there in the future, or if we think we want to add the ability to default some other options that we add (like adding timeout to the struct so it becomes something you an set a default for), then I think ConnectionOptions is fine.

JeremyChiang replied:
After some more thinking and reviewing of all three CBConnectPeripheralOptions available, they are all really semantically closer to "warning options" than "connection options", so I don't really feel comfortable grouping those with timeout, which is definitely a connection option.

If we add something else similar to timeout in the future, then we can add a `ConnectionOptions` struct to group timeout with those new connection options, but even then, I would still prefer to separate them from CBConnectPeripheralOptions/WarningOptions.

nbrooke commented on Bluejay/Bluejay/ConnectionOptions.swift:
```diff
@@ -0,0 +1,33 @@
+import Foundation
+import CoreBluetooth
+
+/// Allow configuring whether iOS can display a system alert when certain conditions are met while your app is suspended, usually an alert dialog outside of your app in the Home screen for example.
```
Comment seems a little overly specific. Even though the only per connection options right now ARE for whether to show the system alert, the struct IN THEORY could specify other per-connection operations, either stuff added to CB later, or something of our own (like the suggestion above to put timeout in this, even though, as I said I'm not sure that's actually a good idea in practice). I think the documentation for the struct itself should just describe the overall purpose of the options, rather than focusing on the specific options it has now.

JeremyChiang replied:
Change to this comment will depend on whether we decide to rename this struct, but either way, yes, I think we can and should focus more on the overall purpose. Once we have a decision, I will improve the comment here as discussed.

JeremyChiang commented:
@nbrooke and @JustusvonBrandt I've addressed and updated all the feedbacks in my last two commits. I've also tested this branch briefly using one of our existing projects that uses Bluejay, and it worked fine. I think this is now ready for merge into branch `v0.7.0` if there are no other major concerns or questions.

In summary, this feature will make clear the differences and allow users to configure all `CBCentralManager` and `CBPeripheralConnection` options should they choose to, but they are not always required (can use Bluejay provided default when the param is not optional, or don't need to specify when the param can be an optional with default). The only breaking API is `start`. `connect` has changed too, but not breaking due to the new params for `connect` being the aforementioned optional default parameters.

All docs and changelog updated as well.

nbrooke commented on Bluejay/Bluejay/Bluejay.swift:
```diff
@@ -129,24 +132,18 @@ public class Bluejay: NSObject {
-        coreBluetoothState: (manager: CBCentralManager, peripheral: CBPeripheral?)? = nil
-        )
-    {
+    public func start(mode: StartMode, connectionObserver observer: ConnectionObserver? = nil) {
```
Should `mode` maybe have a default of `.new(StartOptions.default)`, it's a little unfortunate to HAVE to pass in a start options in the case where just start() would have worked before. 

If it's not possible to do a default parameter pointing at another var like that, maybe there should at least be an alternate version of `start` that just takes a connection observer and calls the main start with the default.

nbrooke commented:
Other than the above minor quibble, this looks good to me.
